File: wasm/sample/src/xplpc/client/remote-client.ts
import { XPLPC } from "../core/xplpc";
import { XCallbackList } from "../data/callback-list";
import { XRequest } from "../message/request";
import { ExceptionMessage } from "../util/exception-message";
import { Log } from "../util/log";
import { UniqueID } from "../util/unique-id";

export class XRemoteClient {
    static call<T>(request: XRequest): Promise<T | undefined> {
        return new Promise<T | undefined>((resolve) => {
            if (!XPLPC.shared().initialized) {
                Log.e("[XRemoteClient : call] The WASM module is not initialized");
                return resolve(undefined);
            }

            try {
                UniqueID.generate().then((key: string) => {
                    XCallbackList.shared().add(key, (response: string) => {
                        return resolve(XPLPC.shared().config.serializer.decodeFunctionReturnValue<T>(response));
                    });

                    XPLPC.shared().module.ProxyClient.callFromJavascript(request.data(), (response: string) => {
                        XCallbackList.shared().execute(key, response);
                    });
                });
            } catch (e: unknown) {
                Log.e("[XRemoteClient : call] Error: " + ExceptionMessage.get(e));
                return resolve(undefined);
            }
        });
    }
}

File: flutter/plugin/lib/client/remote_client.dart
import 'package:ffi/ffi.dart';
import 'package:xplpc/core/xplpc.dart';
import 'package:xplpc/data/callback_list.dart';
import 'package:xplpc/message/request.dart';
import 'package:xplpc/proxy/platform_proxy.dart';
import 'package:xplpc/type/typedefs.dart';
import 'package:xplpc/util/log.dart';
import 'package:xplpc/util/unique_id.dart';

class RemoteClient {
  static void call<T>(Request request, RemoteClientCallback callback) {
    try {
      UniqueID.generate().then((key) {
        CallbackList.instance.add(key, (String response) {
          callback(XPLPC.instance.config.serializer
              .decodeFunctionReturnValue<T>(response));
        });

        var nativeKey = key.toNativeUtf8();
        var nativeData = request.data().toNativeUtf8();

        PlatformProxy.nativeCallProxyFunc(
          nativeKey,
          nativeKey.length,
          nativeData,
          nativeData.length,
        );
      });

      return;
    } on Exception catch (e) {
      Log.e("[RemoteClient : call] Error: $e");
    }

    callback(null);
  }
}
